x-pack/plugins/ml/public/data_frame/components/define_pivot/pivot_preview.tsx

```diff
@@ -32,36 +43,58 @@ export const PivotPreview: React.SFC<Props> = React.memo(({ aggs, groupBy, query
     return null;
   }
 
-  const [loading, setLoading] = useState(false);
-  const [dataFramePreviewData, setDataFramePreviewData] = useState([]);
-
-  useEffect(
-    () => {
-      if (aggs.length === 0) {
-        setDataFramePreviewData([]);
-        return;
-      }
-
-      setLoading(true);
-
-      const request = getDataFramePreviewRequest(indexPattern.title, query, groupBy, aggs);
-
-      ml.dataFrame
-        .getDataFrameTransformsPreview(request)
-        .then((resp: any) => {
-          setDataFramePreviewData(resp.preview);
-          setLoading(false);
-        })
-        .catch((resp: any) => {
-          setDataFramePreviewData([]);
-          setLoading(false);
-        });
-    },
-    [indexPattern.title, aggs, groupBy, query]
+  const { dataFramePreviewData, errorMessage, status } = usePivotPreviewData(
+    indexPattern,
+    query,
+    aggs,
+    groupBy
   );
 
+  if (status === PIVOT_PREVIEW_STATUS.ERROR) {
+    return (
+      <EuiPanel grow={false}>
+        <PreviewTitle />
+        <EuiCallOut
+          title={i18n.translate(
+            'xpack.ml.dataframe.sourceIndexPreview.dataFramePivotPreviewError',
+            {
+              defaultMessage: 'An error occurred loading the pivot preview.',
+            }
+          )}
+          color="danger"
+          iconType="cross"
+        >
+          <p>{errorMessage}</p>
+        </EuiCallOut>
+      </EuiPanel>
+    );
+  }
+
   if (dataFramePreviewData.length === 0) {
-    return null;
+    return (
```

Comment:
If the source index query returns no results, I wonder if it is worth hiding the 'Pivot preview unavailable' callout.

![image](https://user-images.githubusercontent.com/7405507/56653288-4244cf00-6685-11e9-94be-a697d2c44843.png)

Reply:
I'll revisit this in a follow-up!